plugins/azure-devops/src/hooks/useReadme.test.tsx

```diff
@@ -0,0 +1,108 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { renderHook, waitFor } from '@testing-library/react';
+import { Entity } from '@backstage/catalog-model';
+import { Readme } from '@backstage/plugin-azure-devops-common';
+import { TestApiProvider } from '@backstage/test-utils';
+import { AzureDevOpsApi, azureDevOpsApiRef } from '../api';
+import { useReadme } from './useReadme';
+
+describe('useReadme', () => {
+  const azureDevOpsApiMock = {
+    getReadme: jest.fn(),
+  };
+  const azureDevOpsApi =
+    azureDevOpsApiMock as Partial<AzureDevOpsApi> as AzureDevOpsApi;
+
+  const Wrapper = (props: { children?: React.ReactNode }) => (
+    <TestApiProvider apis={[[azureDevOpsApiRef, azureDevOpsApi]]}>
+      {props.children}
+    </TestApiProvider>
+  );
+
+  it('should provide a Readme', async () => {
+    const entity: Entity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Component',
+      metadata: {
+        namespace: 'default',
+        name: 'project-repo',
+        annotations: {
+          'dev.azure.com/project-repo': 'projectName/repoName',
+        },
+      },
+    };
+    const readme: Readme = {
+      url: 'https://dev.azure.com/org/project/repo',
+      content: 'This is some fake README content',
+    };
+    azureDevOpsApiMock.getReadme.mockResolvedValue({
+      item: readme,
+    });
+    const { result } = renderHook(() => useReadme(entity), {
+      wrapper: Wrapper,
+    });
+
+    expect(result.current.loading).toEqual(true);
+
+    await waitFor(() => {
+      expect(result.current.item).toEqual({
+        item: readme,
+      });
+    });
+  });
+
+  it('should return throw when annotation missing', async () => {
+    const entity: Entity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Component',
+      metadata: {
+        namespace: 'default',
+        name: 'project-repo',
+      },
+    };
+
+    expect(() =>
+      renderHook(() => useReadme(entity), {
+        wrapper: Wrapper,
+      }),
+    ).toThrow('Value for annotation "dev.azure.com/project" was not found');
+  });
+
+  it('should return throw when annotation invalid', async () => {
+    const entity: Entity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Component',
+      metadata: {
+        namespace: 'default',
+        name: 'project-repo',
+        annotations: {
+          'dev.azure.com/project-repo': 'fake',
+        },
+      },
+    };
+
+    expect(() =>
+      renderHook(() => useReadme(entity), {
+        wrapper: Wrapper,
+      }),
+    ).toThrow(
```

Comment:
So the hook can fail in two distinct ways: throwing synchronously, or returning an error asynchronously. This goes for the other ones too. That's probably fine if it's intentional (they are in a sense two different classes of error), but I wanted to highlight that it's something that consumers might be surprised by. I know these hooks are for internal use, but that immediate throw is one that's more likely to bubble immediately up to the react root and give a more violent crash than whatever handling is in place for the "calm" return of an Error.

So in summary - might it make sense to move all `getAnnotationValuesFromEntity` calls into the `useAsync` to unify how errors are reported?

Reply:
Yes, that was something I was thinking about, will do that in a follow up PR 👍